sdk/core/src/services/syft/mod.rs

```diff
@@ -0,0 +1,368 @@
+use serde_json::{json, Value};
+use std::env;
+use std::path::{Path, PathBuf};
+use std::process::{Command, Output};
+use thiserror::Error;
+
+const CYCLONEDX_JSON_FORMAT: &str = "cyclonedx-json";
+
+/*
+   =============================================================
+   ==> scheme:type/namespace/name@version?qualifiers#subpath <==
+   =============================================================
+
+   The definition for each components is:
+
+   => scheme:     this is the URL scheme with the constant value of "pkg". One of the primary
+                  reason for this single scheme is to facilitate the future official
+                  registration of the "pkg" scheme for package URLs. Required.
+
+   => type:       the package "type" or package "protocol" such as maven, npm, nuget, gem,
+                  pypi, etc. Required.
+
+   => namespace:  some name prefix such as a Maven groupid, a Docker image owner, a GitHub user or
+                  organization. Optional and type-specific.
+
+   => name:       the name of the package. Required.
+
+   => version:    the version of the package. Optional.
+
+   => qualifiers: extra qualifying data for a package such as an OS, architecture, a distro, etc.
+                  Optional and type-specific.
+
+   => subpath:    extra subpath within a package, relative to the package root. Optional.
+*/
+fn generate_stopgap_purl(
+    full_name: String,
+    commit_hash: String,
+    cataloger: Option<String>,
+    sub_path: Option<String>,
+) -> String {
+    let cataloger = match cataloger {
+        Some(cataloger) => cataloger,
+        None => String::from("no-cataloger"),
+    };
+
+    let purl = format!("pkg:{}:{}@{}", cataloger, full_name, commit_hash);
+
+    match sub_path {
+        None => purl,
+        Some(value) => {
+            if value.is_empty() || value == "/" {
+                purl
+            } else {
+                let mut owned_purl = purl;
+                owned_purl.push_str(format!("#{}", value).as_str());
+                owned_purl
+            }
+        }
+    }
+}
+
+/// Syft Service
+pub(crate) struct Service {
```

Comment:
I was thinking that services could be defined the same way.  Then, in the future, should we need to add a trait to all Harbor services, then we would have them all defined as structs rather than a simple grouping of functions.  Also, it's nice for new developers to think "If I need to use a service from core, then I know what to look for and have a basic understanding of how they are used due to convention".  That's my take.